### PR TITLE
kubeadm: increase ut coverage kubeadm/app/phases/dns

### PR DESCRIPTION
**Pull request by my-git9**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Increase coverage kubeadm/app/phases/dns

Before:

```
        k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns     coverage: 45.7% of statements
```

After this change:

```
        k8s.io/kubernetes/cmd/kubeadm/app/phases/addons/dns     coverage: 72.8% of statements
```

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
NONE
```

### CODE REVIEW

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 1]
```diff
@@ -550,3 +553,230 @@ func createClientAndCoreDNSManifest(t *testing.T, corefile, coreDNSVersion strin
...
+				return
+			}
+			if *got != tt.want {
+				t.Errorf("deployedDNSReplicas() = %v, want %v", *got, tt.want)
+			}
+		})
+	}
+}
+
+func createClient(t *testing.T, replisetsize int32, deploymentsize int) *clientsetfake.Clientset {
```
Comment (pacoxu) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:615]:
```suggestion
// replicas is replica of each DNS deployment
// deploymentSize is the number of deployments with `k8s-app=kube-dns` label.
func newMockClientForTest(t *testing.T, replicas int32, deploymentSize int) *clientsetfake.Clientset {
```

To make the test func name more clear.

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 2]
```diff
@@ -550,3 +553,230 @@ func createClientAndCoreDNSManifest(t *testing.T, corefile, coreDNSVersion strin
...
+		printManifest bool
+	}
+	tests := []struct {
+		name    string
+		args    args
+		wantOut string
+		wantErr bool
+	}{
+		{
+			name: "cfg is invalid",
```
Comment (pacoxu) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:656]: Better mention what is invalid.

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 3]
```diff
@@ -550,3 +553,230 @@ func createClientAndCoreDNSManifest(t *testing.T, corefile, coreDNSVersion strin
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			out := &bytes.Buffer{}
+			if err := coreDNSAddon(tt.args.cfg, tt.args.client, &tt.args.replicas, out, tt.args.printManifest); (err != nil) != tt.wantErr {
+				t.Errorf("coreDNSAddon() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if gotOut := out.String(); gotOut != tt.wantOut {
+				t.Errorf("coreDNSAddon() = %v, wantOut %v", gotOut, tt.wantOut)
```
Comment (pacoxu) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:715]:
```suggestion
				t.Errorf("Actual output of coreDNSAddon() does not match expected.\nActual:  %v\nExpected: %v\n", gotOut, tt.wantOut)
```
As the output is mutliline, better to add `\n` before and after.

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 4]
```diff
@@ -550,3 +553,230 @@ func createClientAndCoreDNSManifest(t *testing.T, corefile, coreDNSVersion strin
...
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			out := &bytes.Buffer{}
+			if err := EnsureDNSAddon(tt.args.cfg, tt.args.client, out, tt.args.printManifest); (err != nil) != tt.wantErr {
+				t.Errorf("EnsureDNSAddon() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if gotOut := out.String(); gotOut != tt.wantOut {
+				t.Errorf("EnsureDNSAddon() = %v, want %v", gotOut, tt.wantOut)
```
Comment (pacoxu) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:778]: ditto.

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 5]
```diff
@@ -550,3 +553,230 @@ func createClientAndCoreDNSManifest(t *testing.T, corefile, coreDNSVersion strin
...
+		},
```
Comment (chendave) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:768]: I am thinking is there any better way to deal with long tailing string, it's really ugly to see. Is there any such sample defined in the code base so you can just use it?

**Review by chendave — COMMENTED** [review]
thanks for the patch!

**Comment by chendave** [comment]
pls retitle the subject as "kubeadm: xxx" to follow the convention of this component.

**Review by pacoxu — APPROVED** [review]
Overall LGTM

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 6]
```diff
@@ -550,3 +553,633 @@ func createClientAndCoreDNSManifest(t *testing.T, corefile, coreDNSVersion strin
...
+			},
+			deploymentsize: 3,
+			want:           5,
+			wantErr:        true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			client := newMockClientForTest(t, tt.args.dnsreplicas, tt.deploymentsize)
+			got, err := deployedDNSReplicas(client, tt.args.replicas)
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:603]: i don't think we need to test variadic number of DNS replicas?
see the code around - deployedDNSReplicas() we always call it with 2, unless i'm mistaken.

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 7]
```diff
@@ -550,3 +553,633 @@ func createClientAndCoreDNSManifest(t *testing.T, corefile, coreDNSVersion strin
...
+			},
+		}, metav1.CreateOptions{})
+		if err != nil {
+			t.Fatalf("error creating deployment: %v", err)
+		}
+	}
+	return client
+}
+
+func Test_coreDNSAddon(t *testing.T) {
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:644]:
```suggestion
func TestCoreDNSAddon(t *testing.T) {
```

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go (hunk 5 above)
Comment (neolit123) [reply cmd/kubeadm/app/phases/addons/dns/dns_test.go:1171]: there is a Dedent library we can use?
see https://github.com/kubernetes/kubernetes/blob/0852a2759afad8dce8f63ed7b5be2859084996d1/cmd/kubeadm/app/cmd/cmd.go#L38

File: cmd/kubeadm/app/phases/addons/dns/dns_test.go [hunk 8]
```diff
@@ -550,3 +553,633 @@ func createClientAndCoreDNSManifest(t *testing.T, corefile, coreDNSVersion strin
...
+			if *got != tt.want {
+				t.Errorf("deployedDNSReplicas() = %v, want %v", *got, tt.want)
+			}
+		})
+	}
+}
+
+// replicas is replica of each DNS deployment
+// deploymentSize is the number of deployments with `k8s-app=kube-dns` label.
+func newMockClientForTest(t *testing.T, replicas int32, deploymentSize int) *clientsetfake.Clientset {
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/addons/dns/dns_test.go:617]: try moving this function at the bottom below all test functions.

**Review by neolit123 — COMMENTED** [review]
+1 for increasing coverage. thanks for the PR.
added some nit around naming, plus some questions around replica counts.

**Review by neolit123 — COMMENTED** [review]
/approve
but i think another reviewer should double check everything.
thanks

**Comment by neolit123** [comment]
/hold
please drop hold after lgtm label

**Pull request merged**